torch_xla/experimental/assume_pure.py

```diff
@@ -37,3 +46,128 @@ def j2t_autograd(fn):
+
+
+def assume_pure_torch(func=None, use_cache=False):
+  """Decorator to mark a function as pure for PyTorch/XLA.
+  This decorator builds an XLA computation from the function and caches it.
+  The decorated function must be pure (i.e. no side-effects, behavior
+  only depends on inputs). 
+  Args:
+    func: The function to be decorated.
+    use_cache: If True, caches the XLA computation for the function with
+      the same name as the function. It is the user's responsibility to ensure
+      that the function is called with the same input shapes and types each time
+      when using this.
+  NOTE: This decorator only works for forward pass.
+  """
+  assert not torch.is_grad_enabled()
+
+  def wrapper(_func):
+
+    @wraps(_func)
+    def inner(*args, **kwargs):
+      global _XLA_COMPUTATION_CACHE
+
+      flat_inputs, input_tree_spec = tree_flatten((args, kwargs))
+      computation_inputs = None
+
+      # TODO: Decide what to include in the cache key.
+      if use_cache and _XLA_COMPUTATION_CACHE.get(_func.__name__,
+                                                  None) is not None:
+        fn_computation, output_tree_spec, hlo_input_id_to_input_index_map, hoisted_vars = _XLA_COMPUTATION_CACHE[
+            _func.__name__]
+        computation_inputs_size = len(hoisted_vars) + len(
+            hlo_input_id_to_input_index_map)
+        computation_inputs = [None] * (computation_inputs_size)
+        for hlo_id, input_index in hlo_input_id_to_input_index_map.items():
+          computation_inputs[hlo_id] = flat_inputs[input_index]
+        for i, var in hoisted_vars.items():
+          computation_inputs[i] = var
+      else:
+        flat_fake_inputs = [make_fake_inputs(a) for a in flat_inputs]
+        fake_args, fake_kwargs = tree_unflatten(flat_fake_inputs,
+                                                input_tree_spec)
+        fake_outputs = _func(*fake_args, **fake_kwargs)
+        flat_fake_outputs, output_tree_spec = tree_flatten(fake_outputs)
+
+        fn_ctx = torch_xla._XLAC.lowering.LoweringContext("FnComputation")
+        fn_ctx.set_name_string("fn_ctx")
+        fn_ctx.build(flat_fake_outputs)
+        fn_hlo = fn_ctx.hlo()
+        fn_computation = xb.computation_from_module_proto(
+            f"xla::xb_computation_{_func.__name__}", fn_hlo)
+
+        computation_inputs, hoisted_vars, hlo_input_id_to_input_index_map = prepare_computation_inputs(
+            fn_ctx, flat_fake_inputs, flat_inputs)
+
+        if use_cache:
+          _XLA_COMPUTATION_CACHE[_func.__name__] = (
+              fn_computation, output_tree_spec, hlo_input_id_to_input_index_map,
+              hoisted_vars)
+
+      result = torch_xla._XLAC._xla_user_computation(
+          f"xla::xb_computation_{_func.__name__}", computation_inputs,
+          fn_computation)
+      result_tree = tree_unflatten(result, output_tree_spec)
+      return result_tree
+
+    return inner
+
+  if func is None:
```

Comment:
remove this bit?  It's a little too clever. 

The UX for passing use_cache can be

```
@functools.partial(
  assume_pure_torch,
  use_cache=True)
def func(...)
```
which is clearer

Reply:
Makes sense. Done!